python/tests/test_writer.py

```diff
@@ -124,59 +124,159 @@ def test_enforce_schema(existing_table: DeltaTable, mode: str):
+
+    for i in range(old_table_data.num_columns):
+        col = old_table_data.schema.field(i)
+        new_data = new_data.add_column(i, col, pa.nulls(new_data.num_rows, col.type))
+
+    old_table_data = old_table_data.append_column(
+        pa.field("new_x", pa.int32()), pa.nulls(old_table_data.num_rows, pa.int32())
+    )
+    old_table_data = old_table_data.append_column(
+        pa.field("new_y", pa.int32()), pa.nulls(old_table_data.num_rows, pa.int32())
+    )
+
+    # define sort order
+    read_data = existing_table.to_pyarrow_table().sort_by(
+        [("utf8", "ascending"), ("new_x", "ascending")]
+    )
+    print(repr(read_data.to_pylist()))
+    concated = pa.concat_tables([old_table_data, new_data])
+    print(repr(concated.to_pylist()))
+    assert read_data == concated
+
+    write_deltalake(existing_table, new_data, mode="overwrite", schema_mode="overwrite")
+
+    assert existing_table.schema().to_pyarrow() == new_data.schema
+
+
+def test_overwrite_schema(existing_table: DeltaTable):
+    new_data_invalid = pa.table(
+        {
+            "utf8": pa.array([1235, 546, 5645]),
+            "new_x": pa.array([1, 2, 3], pa.int32()),
+            "new_y": pa.array([1, 2, 3], pa.int32()),
+        }
+    )
 
     with pytest.raises(DeltaError):
         write_deltalake(
             existing_table,
-            new_data,
+            new_data_invalid,
             mode="append",
-            overwrite_schema=True,
+            schema_mode="overwrite",
             engine="rust",
         )
+
+    new_data = pa.table(
+        {
+            "utf8": pa.array(["bla", "bli", "blubb"]),
+            "new_x": pa.array([1, 2, 3], pa.int32()),
+            "new_y": pa.array([1, 2, 3], pa.int32()),
+        }
+    )
     with pytest.raises(DeltaError):
         write_deltalake(
             existing_table,
             new_data,
-            mode="overwrite",
-            overwrite_schema=False,
+            mode="append",
+            schema_mode="overwrite",
+            engine="rust",
+        )
+
+    write_deltalake(existing_table, new_data, mode="overwrite", schema_mode="overwrite")
+
+    assert existing_table.schema().to_pyarrow() == new_data.schema
+
+
+def test_update_schema_rust_writer_append(existing_table: DeltaTable):
+    with pytest.raises(ValueError):
+        # It's illegal to do schema drift without correct schema_mode
+        write_deltalake(
+            existing_table,
+            pa.table({"x4": pa.array([1, 2, 3])}),
+            mode="append",
+            schema_mode=None,
             engine="rust",
         )
     with pytest.raises(DeltaError):
         write_deltalake(
             existing_table,
-            new_data,
+            pa.table({"x1": pa.array([1, 2, 3])}),
             mode="append",
-            overwrite_schema=False,
+            schema_mode="overwrite",
+            engine="rust",
+        )
+    with pytest.raises(ValueError):
```

Comment:
Can you make the test a bit more explicit by also catching the message or in a seperate test. 

Because this fails because UTF8 column can't be merged since LHS and RHS have different dtypes